datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -655,6 +658,20 @@ impl TreeNodeVisitor for ExprIdentifierVisitor<'_> {
     }
 }
 
+/// Check if the expression is short-circuit expression
+fn is_short_circuit_expression(expr: &Expr) -> bool {
```

Comment:
I think we should put this method on expr so it is easier to find and potentially use 

```rust
impl Expr {
...
  /// returns true if some of this `exprs` subexpressions may not be evaluated
  /// and thus any side effects (like divide by zero) may not be encountered
  fn short_circuits(&self) -> bool { 
   ...
   }
```

Reply:
Additionally, we'd better enumerate all expressions here.
This means we would need to consider whether it's short-circuiting when adding a new expression.